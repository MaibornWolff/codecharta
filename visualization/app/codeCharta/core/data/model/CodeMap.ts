export interface CodeMapNode {
    name: string;
    type: string;
    children?: CodeMapNode[];
    attributes: {
        [key: string]: number
    };
    deltas?: {
        [key: string]: number
    };
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
    markingColor?: string;
}

export interface CodeMap {
    fileName: string;
    apiVersion?: string;
    projectName: string;
    root: CodeMapNode;
    edges?: Edge[];
    attributeTypes?: AttributeTypes;
    blacklist?: Array<BlacklistItem>;
}

export interface Edge {
    fromNodeName: string;
    toNodeName: string;
    attributes: {
        [key: string]: number
    };
    visible?: boolean;
}

export interface AttributeTypes {
    nodes?: {
        [key: string]: AttributeType
    };
    edges?: {
        [key: string]: AttributeType
    };
}

export enum AttributeType {
    absolute, relative
}

export interface BlacklistItem {
    path: string;
    type: BlacklistType;
}

export enum BlacklistType {
    hide = "hide",
    exclude = "exclude"
}