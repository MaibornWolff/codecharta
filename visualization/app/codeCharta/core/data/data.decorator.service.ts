"use strict";

import * as d3 from "d3";
import {CodeMap, CodeMapNode} from "./model/CodeMap";
import {HierarchyNode} from "d3-hierarchy";

/**
 * Decorates the data structure with artificial metrics
 */
export class DataDecoratorService {

    /* @ngInject */
    constructor() {

    }

    /**
     * Decorates the map with the unary metric. This metric is always 1 to allow the same area on all buildings.
     * @param {CodeMap} map
     */
    public decorateMapWithUnaryMetric(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            let descendants: HierarchyNode<CodeMapNode>[] = root.descendants();

            for (let j = 0; j < descendants.length; j++) {
                if (!descendants[j].data.attributes) {
                    descendants[j].data.attributes = {};
                }
                Object.assign(descendants[j].data.attributes, {unary: 1});
            }

        }
    }

    public decorateMapWithVisibleAttribute(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                node.data.visible = true;
            });

        }
    }

    public decorateMapWithOriginAttribute(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                node.data.origin = map.fileName;
            });

        }
    }

    public decorateMapWithPathAttribute(map: CodeMap) {
        if (map && map.root) {
            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                let path = root.path(node);
                let pathString = "";
                path.forEach((pnode) => {
                    pathString += "/" + pnode.data.name;
                });
                node.data.path = pathString;
            });
        }
    }

    public decorateLeavesWithMissingMetrics(maps: CodeMap[], metrics: string[]) {

        maps.forEach((map) => {
            if (map && map.root) {

                let root = d3.hierarchy<CodeMapNode>(map.root);
                root.leaves().forEach((node) => {

                    //make sure attributes exist
                    this.createAttributesIfNecessary(node);

                    //create Metrics
                    for (let i = 0; i < metrics.length; i++) {
                        let metric = metrics[i];
                        if (!node.data.attributes.hasOwnProperty(metric)) {
                            node.data.attributes[metric] = 0;
                        }
                    }

                });

            }
        });

    }

    public decorateParentNodesWithMeanAttributesOfChildren(maps: CodeMap[], metrics: string[]) {

        maps.forEach((map) => {
            if (map && map.root) {

                let root = d3.hierarchy<CodeMapNode>(map.root);
                root.each((node) => {
                    this.decorateNodeWithChildrenMeanMetrics(node, metrics);

                });

            }
        });
    }

    public decorateNodeWithChildrenMeanMetrics(node, metrics: string[]) {
        //make sure attributes exist
        this.createAttributesIfNecessary(node);

        //if attributes is empty define property for each possible metric as a mean function of child metrics
        for (let i = 0; i < metrics.length; i++) {
            let metric = metrics[i];
            if (!node.data.attributes.hasOwnProperty(metric) && node.data.children && node.data.children.length > 0) {
                this.defineAttributeAsMeanMethod(node, metric);
            }
        }

    }

    private defineAttributeAsMeanMethod(node, metric: string) {
        Object.defineProperty(node.data.attributes, metric, {
            enumerable: true,
            get: function () {
                let count = 0;
                let sum = 0;
                let l = node.leaves();
                for (; count < l.length; count++) {
                    sum += l[count].data.attributes[metric];
                }
                return sum / count;
            }
        });
    }

    private createAttributesIfNecessary(node) {
        if (!node.data.attributes) {
            node.data.attributes = {};
        }
    }
}
