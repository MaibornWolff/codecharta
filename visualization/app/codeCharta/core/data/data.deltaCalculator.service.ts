"use strict";

import * as d3 from "d3";
import {CodeMap, CodeMapNode} from "./model/CodeMap";
import {HierarchyNode} from "d3-hierarchy";
import * as deepcopy from "deepcopy";

interface KVObject {
    [key: string]: number
}

/**
 * Calculates the deltas between given maps and modifies the data structure
 */
export class DeltaCalculatorService {

    /* @ngInject */
    constructor() {

    }

    public decorateMapsWithDeltas(firstMap: CodeMap, secondMap: CodeMap) {

        if (firstMap && secondMap && firstMap.root && secondMap.root) {
            let firstRoot = d3.hierarchy<CodeMapNode>(firstMap.root);
            let firstLeaves: HierarchyNode<CodeMapNode>[] = firstRoot.leaves();
            let secondRoot = d3.hierarchy(secondMap.root);
            let secondLeaves: HierarchyNode<CodeMapNode>[] = secondRoot.leaves();

            for (var j = 0; j < firstLeaves.length; j++) {
                for (var k = 0; k < secondLeaves.length; k++) {

                    let fl: HierarchyNode<CodeMapNode> = firstLeaves[j];
                    let sl: HierarchyNode<CodeMapNode> = secondLeaves[k];

                    if (fl.data.name === sl.data.name) {
                        //calculate delta for those nodes attributes and push it to the second leave
                        let firstDeltas = this.calculateAttributeListDelta(sl.data.attributes, fl.data.attributes);
                        let secondDeltas = this.calculateAttributeListDelta(fl.data.attributes, sl.data.attributes);

                        firstLeaves[j].data.deltas = firstDeltas;
                        secondLeaves[k].data.deltas = secondDeltas;

                    }
                }
            }

        }

    }

    /**
     * Unifies both maps and inserts unique nodes into the other map. The inserted nodes have all metrics set to zero.
     * This will calculate deltas correctly between versions and add "empty" buildings to ensure all revisions have an
     * equal building pool. These empty buildings should be viewed with the unary metric as area
     * @param {CodeMap} leftMap
     * @param {CodeMap} rightMap
     * @returns {{leftMap: CodeMap; rightMap: CodeMap}}
     */
    public fillMapsWithNonExistingNodesFromOtherMap(leftMap: CodeMap, rightMap: CodeMap): { leftMap: CodeMap; rightMap: CodeMap; } {
        if (leftMap && rightMap && leftMap.root && rightMap.root) {

            let leftMapCopy: CodeMap = deepcopy.default(leftMap);
            let rightMapCopy: CodeMap = deepcopy.default(rightMap);

            let leftRoot = d3.hierarchy<CodeMapNode>(leftMapCopy.root);
            let rightRoot = d3.hierarchy<CodeMapNode>(rightMapCopy.root);

            this.insertLeftIntoRightWithZeroMetrics(leftRoot, rightRoot);
            this.insertLeftIntoRightWithZeroMetrics(rightRoot, leftRoot);

            return {leftMap: leftMapCopy, rightMap: rightMapCopy};

        } else {
            return {leftMap: leftMap, rightMap: rightMap};
        }
    }

    public removeCrossOriginNodes(map: CodeMap): CodeMap {
        if (map && map.root) {

            let mapCopy: CodeMap = deepcopy.default(map);

            let mapRoot = d3.hierarchy<CodeMapNode>(mapCopy.root);
            mapRoot.each((node)=>{
                if(node.data.children) {
                    node.data.children = node.data.children.filter(x => (x.origin === mapCopy.fileName));
                }
            });

            return mapCopy;

        } else {
            return map;
        }
    }

    private insertLeftIntoRightWithZeroMetrics(left: HierarchyNode<CodeMapNode>, right: HierarchyNode<CodeMapNode>) {

        // are these nodes mergable (same name and have children)?
        if (left.data.name === right.data.name && left.children != undefined && right.children != undefined && left.children.length > 0 && right.children.length > 0) {

            left.children.forEach((leftChild) => {

                let leftChildExistsInRight = this.childExistsInHierarchy(right, leftChild);

                //if left child does not exist in the right node's children, insert it as a new valueless node into rights children
                if (!leftChildExistsInRight) {
                    let copy = this.deepcopy(leftChild);
                    right.data.children.push(copy.data);
                } else {
                    // if left child exists in right nodes children, skip, since we only look for direct children
                    right.children.forEach((rightChildInner) => {
                        left.children.forEach((leftChildInner) => {
                            //step in with pre-filtering
                            if (leftChildInner.data.name !== leftChild.data.name) {
                                if (leftChildInner.data.name === rightChildInner.data.name && leftChildInner.children != undefined && rightChildInner.children != undefined && leftChildInner.children.length > 0 && rightChildInner.children.length > 0) {
                                    this.insertLeftIntoRightWithZeroMetrics(leftChildInner, rightChildInner);
                                }
                            }
                        });
                    });
                }

            });

        }
    }

    private childExistsInHierarchy(hierarchy: HierarchyNode<CodeMapNode>, child: HierarchyNode<CodeMapNode>): boolean {
        let tmp = false;
        hierarchy.children.forEach((rightChild) => {
            if (child.data.name === rightChild.data.name) {
                tmp = true;
            }
        });
        return tmp;
    }

    private deepcopy(nodes: HierarchyNode<CodeMapNode>): HierarchyNode<CodeMapNode> {

        //deepcopy
        let copy: HierarchyNode<CodeMapNode> = deepcopy.default(nodes.copy()); //Hm this seems to be doing the right thing. First shallow copy then a deep copy ?!

        //make own attributes 0
        for (let property in copy.data.attributes) {
            if (copy.data.attributes.hasOwnProperty(property)) {
                copy.data.attributes[property] = 0;
            }
        }

        ////make all ancestors attributes 0
        copy.each((node) => {
            for (var property in node.data.attributes) {
                if (node.data.attributes.hasOwnProperty(property)) {
                    node.data.attributes[property] = 0;
                }
            }
        });

        return copy;

    }

    calculateAttributeListDelta(first: KVObject, second: KVObject) {
        let deltas = {};
        for (var key in second) {
            if (key) {
                let firstValue = first[key] ? first[key] : 0; //assume zero if no value in first
                let secondValue = second[key];
                let delta = secondValue - firstValue;
                deltas[key] = delta;
            }
        }
        return deltas;
    }

}
