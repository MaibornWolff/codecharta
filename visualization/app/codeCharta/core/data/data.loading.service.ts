"use strict";
import {CodeMap} from "./model/CodeMap";
import {DataValidatorService} from "./data.validator.service";
import {DataService} from "./data.service";

/**
 * This service loads maps into the DataService
 */
export class DataLoadingService {

    /* @ngInject */
    constructor(private dataValidatorService: DataValidatorService, private dataService: DataService){

    }

    /**
     * Validates and loads a map from the given file content into the dataService
     * @param {Object} fileContent well formed (schema.json) fileContent
     * @param {Number} revision the revision id
     * @returns {Promise} which resolves when the filecontent is valid and stored in dataService.
     * The Promise rejects when errors happen. The errors are provided as parameters of the rejection function
     */
    loadMapFromFileContent(fileName: string, fileContent: any, revision: number = 0): Promise<CodeMap> {

        return new Promise((resolve, reject) => {

            this.dataValidatorService.validate(fileContent).then(

                ()=>{
                    const map: CodeMap = {
                        fileName: fileName, 
                        projectName: fileContent.projectName,
                        root: fileContent.nodes[0],
                        dependencies: fileContent.dependencies || {}
                    };
                    this.dataService.setMap(map, revision);

                    resolve(map);

                }, (errors) => {
                    reject(errors);
                }

            );

        });

    }


    
}
