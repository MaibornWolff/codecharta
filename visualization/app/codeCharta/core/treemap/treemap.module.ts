"use strict"

import angular from "angular"

import "../data/data.module"
import { TreeMapService } from "./treemap.service"

angular.module("app.codeCharta.core.treemap", ["app.codeCharta.core.data"])

angular.module("app.codeCharta.core.treemap").service("treeMapService", TreeMapService)
