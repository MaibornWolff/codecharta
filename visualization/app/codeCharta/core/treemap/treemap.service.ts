import {CodeMapNode, Edge, BlacklistItem, BlacklistType} from "../data/model/CodeMap";
import {Node} from "../../ui/codeMap/rendering/node";
import {DataService} from "../data/data.service";
import * as d3 from "d3";
import {hierarchy, HierarchyNode} from "d3";
import {TreeMapUtils} from "./treemap.util";
import {CodeMapUtilService} from "../../ui/codeMap/codeMap.util.service";
import {MarkedPackage} from "../settings/settings.service";

export interface ValuedCodeMapNode {
    data: CodeMapNode;
    children?: ValuedCodeMapNode[];
    value: number;
}

export interface SquarifiedValuedCodeMapNode {
    data: CodeMapNode;
    children?: SquarifiedValuedCodeMapNode[];
    parent?: SquarifiedValuedCodeMapNode;
    value: number;
    x0: number;
    y0: number;
    x1: number;
    y1: number;
}

export interface TreeMapSettings {
    size: number;
    areaKey: string;
    heightKey: string;
    margin: number;
    invertHeight: boolean;
    visibleEdges: Edge[];
    searchedNodePaths: string[];
    blacklist: Array<BlacklistItem>;
    fileName: string;
    searchPattern: string;
    hideFlatBuildings: boolean;
    markedPackages: MarkedPackage[]
}

export class TreeMapService {

    private static HEIGHT_DIVISOR = 1;
    private static FOLDER_HEIGHT = 2;
    private static MIN_BUILDING_HEIGHT = 2;
    private static HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND = 0;
    private static PADDING_SCALING_FACTOR = 0.4;

    /* @ngInject */
    constructor(private dataService: DataService) {}

    public createTreemapNodes(data: CodeMapNode, s: TreeMapSettings, edges: Edge[]): Node {
        const squarified: SquarifiedValuedCodeMapNode = this.squarify(data, s, edges);
        const heighted = this.addMapScaledHeightDimensionAndFinalizeFromRoot(squarified, s);
        return heighted;
    }

    public setVisibilityOfNodeAndDescendants(node: CodeMapNode, visibility: boolean) {
        node.visible = visibility;
        hierarchy<CodeMapNode>(node).descendants().forEach((hierarchyNode) => {
            hierarchyNode.data.visible = visibility;
        });
        return node;
    }

    private squarify(data: CodeMapNode, s: TreeMapSettings, edges: Edge[]): SquarifiedValuedCodeMapNode {
        let nodes: HierarchyNode<CodeMapNode> = d3.hierarchy<CodeMapNode>(data);
        const blacklisted = CodeMapUtilService.numberOfBlacklistedNodes(nodes.descendants().map(d=>d.data), s.blacklist);
        let nodesPerSide = 2 * Math.sqrt(nodes.descendants().length - blacklisted);
        let treeMap = d3.treemap<CodeMapNode>()
            .size([s.size + nodesPerSide*s.margin, s.size + nodesPerSide*s.margin])
            .paddingOuter(s.margin * TreeMapService.PADDING_SCALING_FACTOR || 1)
            .paddingInner(s.margin * TreeMapService.PADDING_SCALING_FACTOR || 1);

        return treeMap(nodes.sum((node) => this.calculateValue(node, edges, s))) as SquarifiedValuedCodeMapNode;
    }

    private addMapScaledHeightDimensionAndFinalizeFromRoot(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings): Node {
        const maxHeight = this.dataService.getMaxMetricInAllRevisions(s.heightKey);
        const heightScale = s.size / TreeMapService.HEIGHT_DIVISOR / maxHeight;
        return this.addHeightDimensionAndFinalize(squaredNode, s, heightScale, maxHeight);
    }

    private addHeightDimensionAndFinalize(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings, heightScale: number, maxHeight: number, depth = 0, parent: Node = null): Node {

        let attr = squaredNode.data.attributes || {};
        let heightValue = attr[s.heightKey];

        if (heightValue === undefined || heightValue === null) {
            heightValue = TreeMapService.HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND;
        }

        if (CodeMapUtilService.isBlacklisted(squaredNode.data, s.blacklist, BlacklistType.hide)) {
            squaredNode.data = this.setVisibilityOfNodeAndDescendants(squaredNode.data, false);
        }

        const finalNode = TreeMapUtils.buildNodeFrom(squaredNode, heightScale, heightValue, maxHeight, depth, parent, s, TreeMapService.MIN_BUILDING_HEIGHT, TreeMapService.FOLDER_HEIGHT);

        if (squaredNode.children && squaredNode.children.length > 0) {
            const finalChildren: Node[] = [];
            for (let i = 0; i < squaredNode.children.length; i++) {
                finalChildren.push(this.addHeightDimensionAndFinalize(squaredNode.children[i], s, heightScale, maxHeight, depth + 1, finalNode));
            }
            finalNode.children = finalChildren;
        }
        return finalNode;

    }

    private isDeletedNodeFromComparisonMap(node: CodeMapNode, s: TreeMapSettings): boolean {
        return node &&
                node.deltas &&
                node.origin !== s.fileName &&
                node.deltas[s.heightKey] < 0 &&
                node.attributes[s.areaKey] === 0;
    }

    private calculateValue(node: CodeMapNode, edges: Edge[], s: TreeMapSettings): number {

        let result = 0;

        if(CodeMapUtilService.isBlacklisted(node, s.blacklist, BlacklistType.exclude)) {
            return 0;
        }

        if(this.isDeletedNodeFromComparisonMap(node, s)) {
            return Math.abs(node.deltas[s.areaKey]);
        }

        if ((!node.children || node.children.length === 0)) {
            if(node.attributes && node.attributes[s.areaKey]) {
                result = node.attributes[s.areaKey] || 0;
            } else {
                result = this.getEdgeValue(node, edges, s.areaKey);
            }
        }
        return result;
    }

    private getEdgeValue(node: CodeMapNode, edges: Edge[], key: string) {
        let filteredEdgeAttributes: number[] = [];

        if (edges) {
            edges.forEach((edge)=> {
                if (edge.fromNodeName == node.path || edge.toNodeName == node.path) {
                    filteredEdgeAttributes.push(edge.attributes[key]);
                }
            });
        }

        if (filteredEdgeAttributes) {
            return filteredEdgeAttributes.sort().reverse()[0];
        }
        return 1;
    }
}

