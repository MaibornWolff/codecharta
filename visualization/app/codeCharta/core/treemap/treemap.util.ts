import {SquarifiedValuedCodeMapNode, TreeMapSettings} from "./treemap.service";
import {node} from "../../ui/codeMap/rendering/node";

export class TreeMapUtils {

    public static countNodes(node: { children?: any }): number {
        let count = 1;
        if (node.children && node.children.length > 0) {
            for (let i = 0; i < node.children.length; i++) {
                count += TreeMapUtils.countNodes(node.children[i]);
            }
        }
        return count;
    }

    public static isNodeLeaf(node: { children?: any }): boolean {
        return !(node.children && node.children.length > 0);
    }

    public static buildNodeFrom(squaredNode: SquarifiedValuedCodeMapNode,
                                heightScale: number,
                                heightValue: number,
                                maxHeight: number,
                                depth: number,
                                parent: node,
                                s: TreeMapSettings,
                                minHeight: number,
                                folderHeight: number): node {

        if (s.invertHeight) {
            heightValue = (maxHeight - heightValue);
        }

        const flattened = TreeMapUtils.isNodeToBeFlat(squaredNode, s);
        if (flattened) {
            heightValue = minHeight;
        }

        return {
            name: squaredNode.data.name,
            width: squaredNode.x1 - squaredNode.x0,
            height: Math.abs(TreeMapUtils.isNodeLeaf(squaredNode) ? Math.max(heightScale * heightValue, minHeight) : folderHeight),
            length: squaredNode.y1 - squaredNode.y0,
            depth: depth,
            x0: squaredNode.x0,
            z0: depth * folderHeight,
            y0: squaredNode.y0,
            isLeaf: TreeMapUtils.isNodeLeaf(squaredNode),
            attributes: squaredNode.data.attributes,
            deltas: squaredNode.data.deltas,
            parent: parent,
            heightDelta: squaredNode.data.deltas && squaredNode.data.deltas[s.heightKey] ? heightScale * squaredNode.data.deltas[s.heightKey] : 0,
            visible: squaredNode.data.visible && !(TreeMapUtils.isNodeLeaf(squaredNode) && s.hideFlatBuildings && flattened),
            path: squaredNode.data.path,
            origin: squaredNode.data.origin,
            link: squaredNode.data.link,
            children: [],
            markingColor: parseInt(squaredNode.data.markingColor),
            flat: flattened,
        };

    }

    private static isNodeToBeFlat(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings): boolean {
        let flattened = false;
        if (s.visibleEdges && s.visibleEdges.length > 0) {
            flattened = this.nodeHasNoVisibleEdges(squaredNode, s);
        }

        if (s.searchedNodePaths && s.searchPattern && s.searchPattern.length > 0) {
            flattened = (s.searchedNodePaths.length == 0) ? true : this.isNodeNonSearched(squaredNode, s);
        }
        return flattened;
    }

    private static nodeHasNoVisibleEdges(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings): boolean {
        return s.visibleEdges.filter(edge =>
            squaredNode.data.path === edge.fromNodeName ||
            squaredNode.data.path === edge.toNodeName).length == 0;
    }

    private static isNodeNonSearched(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings): boolean {
        return s.searchedNodePaths.filter(path =>
            path == squaredNode.data.path).length == 0;
    }
}