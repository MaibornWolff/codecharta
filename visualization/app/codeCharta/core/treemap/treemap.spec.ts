import {TreeMapService, TreeMapSettings} from "./treemap.service";
import {CodeMapNode} from "../data/model/CodeMap";
import {node} from "../../ui/codeMap/rendering/node";
import {DataService} from "../data/data.service";

describe("treemap service", ()=> {

    let treeMapService: TreeMapService;
    let dataServiceMock: DataService;

    beforeEach(()=>{
        dataServiceMock = {
            getMaxMetricInAllRevisions: jest.fn()
        };
        dataServiceMock.getMaxMetricInAllRevisions.mockReturnValue(1000);
        treeMapService = new TreeMapService(dataServiceMock);
    });

    function buildSampeTreemapSettings(): TreeMapSettings {
        return {
            size: 200,
            areaKey: "mcc",
            heightKey: "fun",
            margin: 0,
            invertHeight: false,
            visibleEdges: [],
            searchedNodePaths: [],
            blacklist:[],
            fileName:"currentFile.json"
        };
    }

    function buildSampleAttributes() {
        return {
            "mcc": 1,
            "fun": 10,
            "lin": 100,
        };
    }

    function buildSimpleHierarchy() {
        let root: CodeMapNode = {
            name: "root",
            path: "/root",
            attributes: {},
            type: "Folder"
        };
        let firstChild: CodeMapNode = {
            name: "firstChild",
            path: root.path + "/firstChild",
            attributes: {},
            type: "File"
        };
        let secondChild: CodeMapNode = {
            name: "secondChild",
            path: root.path + "/secondChild",
            attributes: {},
            type: "File"
        };
        root.children = [firstChild, secondChild];
        return { root, firstChild, secondChild };
    }

    function buildComplexHierarchy() {
        let { root, firstChild, secondChild } = buildSimpleHierarchy();
        firstChild.type = "Folder"
        let firstGrandChild = { 
            name: "firstGrandChild", 
            path: firstChild.path+"/firstGrandChild", 
            attributes: root.attributes,
            type: "File"
        };

        let secondGrandChild = { 
            name: "secondGrandChild", 
            path: firstChild.path+"/secondGrandChild", 
            attributes: root.attributes,
            type: "File" 
        };

        firstChild.children = [firstGrandChild, secondGrandChild];
        return { root, firstChild, secondChild , firstGrandChild, secondGrandChild};
    }

    it("only root node", ()=> {
        let { root } = buildSimpleHierarchy();
        root.attributes = buildSampleAttributes();
        root.children = [];
        let node: node = treeMapService.createTreemapNodes(root, buildSampeTreemapSettings(), null);
        expect(node).toMatchSnapshot();
    });

    it("root node with two direct children", ()=> {
        let { root, firstChild, secondChild } = buildSimpleHierarchy();
        root.attributes = firstChild.attributes = secondChild.attributes = buildSampleAttributes();
        let node: node = treeMapService.createTreemapNodes(root, buildSampeTreemapSettings(), null);
        expect(node).toMatchSnapshot();
    });

    it("root node with two direct children and some grand children", ()=> {
        let { root, firstChild, secondChild, firstGrandChild, secondGrandChild } = buildComplexHierarchy();
        root.attributes = firstChild.attributes = secondChild.attributes = firstGrandChild.attributes = secondGrandChild.attributes = buildSampleAttributes();
        let node: node = treeMapService.createTreemapNodes(root, buildSampeTreemapSettings(), null);
        expect(node).toMatchSnapshot();
    });

    describe("CodeMap value calculation", ()=> {

        it("if a node was deleted from previous revision it should still be visible and have positive width/length", ()=> {
            
            // given map with one node deleted in comparison to previous revision
            let root = buildSimpleHierarchy().root;
            root.attributes = {"myArea": 22, "myHeight": 12}
            root.deltas = {};
            root.children[0].attributes = {"myArea": 44, "myHeight": 63}
            root.children[0].deltas = {"myArea": 20, "myHeight": 0}
            root.children[0].origin = "file.json";
            root.children[1].attributes = {"myArea": 0, "myHeight": 0}
            root.children[1].deltas = {"myArea": -40, "myHeight": -80}
            root.children[1].origin = "notfile.json";

            // given settings
            let settings = buildSampeTreemapSettings();
            settings.fileName = "file.json";
            settings.areaKey = "myArea";
            settings.heightKey = "myHeight";
            settings.size=1000;
            
            // when
            let node: node = treeMapService.createTreemapNodes(root, settings, null);

            //then
            expect(node.children[1].name).toBe("secondChild");
            expect(node.children[1].width).toBeGreaterThan(0);
            expect(node.children[1].length).toBeGreaterThan(0);

        });

        it("attribute exists, no children", ()=> {
            let { root } = buildSimpleHierarchy();
            root.children = [];
            root.attributes = {a:100};
            let node: node = treeMapService.createTreemapNodes(root, buildSampeTreemapSettings(), null);
            expect(node.attributes["a"]).toBe(100);
        });

        it("attribute do not exists, no children", ()=> {
            let { root } = buildSimpleHierarchy();
            root.children = [];
            let node: node = treeMapService.createTreemapNodes(root, buildSampeTreemapSettings(), null);
            expect(node.attributes["b"]).toBe(undefined);
        });

        it("attribute do not exists, multiple children with non existant attributes", ()=> {
            let { root  } = buildSimpleHierarchy();
            let treemapSettings = buildSampeTreemapSettings();
            treemapSettings.heightKey = "b";
            treemapSettings.areaKey = "b";
            let node: node = treeMapService.createTreemapNodes(root, treemapSettings, null);
            expect(node.attributes["b"]).toBe(undefined);
        });

        it("attribute do not exists, multiple children with existant attributes. should be undefined since it is not the job of the treemap service to add non existant attributes", ()=> {
            let { root } = buildSimpleHierarchy();
            expect(root.attributes["a"]).toBe(undefined);
        });

    });


});

