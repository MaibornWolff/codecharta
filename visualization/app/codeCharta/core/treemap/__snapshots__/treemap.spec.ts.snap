// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treemap service only root node 1`] = `
Object {
  "attributes": Object {
    "fun": 10,
    "lin": 100,
    "mcc": 1,
  },
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 2,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 200,
  "link": undefined,
  "markingColor": NaN,
  "name": "root",
  "origin": undefined,
  "parent": null,
  "path": "/root",
  "visible": undefined,
  "width": 200,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap service root node with two direct children 1`] = `
Object {
  "attributes": Object {
    "fun": 10,
    "lin": 100,
    "mcc": 1,
  },
  "children": Array [
    Object {
      "attributes": Object {
        "fun": 10,
        "lin": 100,
        "mcc": 1,
      },
      "children": Array [],
      "deltas": undefined,
      "depth": 1,
      "flat": false,
      "height": 2,
      "heightDelta": 0,
      "isLeaf": true,
      "length": 98.5,
      "link": undefined,
      "markingColor": NaN,
      "name": "firstChild",
      "origin": undefined,
      "parent": [Circular],
      "path": "/root/firstChild",
      "visible": undefined,
      "width": 198,
      "x0": 1,
      "y0": 1,
      "z0": 2,
    },
    Object {
      "attributes": Object {
        "fun": 10,
        "lin": 100,
        "mcc": 1,
      },
      "children": Array [],
      "deltas": undefined,
      "depth": 1,
      "flat": false,
      "height": 2,
      "heightDelta": 0,
      "isLeaf": true,
      "length": 98.5,
      "link": undefined,
      "markingColor": NaN,
      "name": "secondChild",
      "origin": undefined,
      "parent": [Circular],
      "path": "/root/secondChild",
      "visible": undefined,
      "width": 198,
      "x0": 1,
      "y0": 100.5,
      "z0": 2,
    },
  ],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 2,
  "heightDelta": 0,
  "isLeaf": false,
  "length": 200,
  "link": undefined,
  "markingColor": NaN,
  "name": "root",
  "origin": undefined,
  "parent": null,
  "path": "/root",
  "visible": undefined,
  "width": 200,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap service root node with two direct children and some grand children 1`] = `
Object {
  "attributes": Object {
    "fun": 10,
    "lin": 100,
    "mcc": 1,
  },
  "children": Array [
    Object {
      "attributes": Object {
        "fun": 10,
        "lin": 100,
        "mcc": 1,
      },
      "children": Array [
        Object {
          "attributes": Object {
            "fun": 10,
            "lin": 100,
            "mcc": 1,
          },
          "children": Array [],
          "deltas": undefined,
          "depth": 2,
          "flat": false,
          "height": 2,
          "heightDelta": 0,
          "isLeaf": true,
          "length": 64.33333333333333,
          "link": undefined,
          "markingColor": NaN,
          "name": "firstGrandChild",
          "origin": undefined,
          "parent": [Circular],
          "path": "/root/firstChild/firstGrandChild",
          "visible": undefined,
          "width": 196,
          "x0": 2,
          "y0": 2,
          "z0": 4,
        },
        Object {
          "attributes": Object {
            "fun": 10,
            "lin": 100,
            "mcc": 1,
          },
          "children": Array [],
          "deltas": undefined,
          "depth": 2,
          "flat": false,
          "height": 2,
          "heightDelta": 0,
          "isLeaf": true,
          "length": 64.33333333333333,
          "link": undefined,
          "markingColor": NaN,
          "name": "secondGrandChild",
          "origin": undefined,
          "parent": [Circular],
          "path": "/root/firstChild/secondGrandChild",
          "visible": undefined,
          "width": 196,
          "x0": 2,
          "y0": 67.33333333333333,
          "z0": 4,
        },
      ],
      "deltas": undefined,
      "depth": 1,
      "flat": false,
      "height": 2,
      "heightDelta": 0,
      "isLeaf": false,
      "length": 131.66666666666666,
      "link": undefined,
      "markingColor": NaN,
      "name": "firstChild",
      "origin": undefined,
      "parent": [Circular],
      "path": "/root/firstChild",
      "visible": undefined,
      "width": 198,
      "x0": 1,
      "y0": 1,
      "z0": 2,
    },
    Object {
      "attributes": Object {
        "fun": 10,
        "lin": 100,
        "mcc": 1,
      },
      "children": Array [],
      "deltas": undefined,
      "depth": 1,
      "flat": false,
      "height": 2,
      "heightDelta": 0,
      "isLeaf": true,
      "length": 65.33333333333334,
      "link": undefined,
      "markingColor": NaN,
      "name": "secondChild",
      "origin": undefined,
      "parent": [Circular],
      "path": "/root/secondChild",
      "visible": undefined,
      "width": 198,
      "x0": 1,
      "y0": 133.66666666666666,
      "z0": 2,
    },
  ],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 2,
  "heightDelta": 0,
  "isLeaf": false,
  "length": 200,
  "link": undefined,
  "markingColor": NaN,
  "name": "root",
  "origin": undefined,
  "parent": null,
  "path": "/root",
  "visible": undefined,
  "width": 200,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;
